Adds Playwright, first tests and linting.

tharsheblows opened the pull request:
Fixes #1528.

This runs in your local database and is noisy! I'd rather run it with a separate database but that might be overkill, I don't know. Currently it causes the lint action to fail as there's not the executables installed.

delawski reviewed (approved):
@tharsheblows Looks great so far!

I think we could use this setup to run E2E tests locally for now and not worry about the CI integration at this point. If you think that's a good approach, we just need to make sure `test:e2e` is not run in GitHub action.

delawski commented on .gitignore:
<!-- general note -->
**📝 note:** @tharsheblows Let's also make sure `.distignore` is kept up to date.

tharsheblows replied:
Thank you! Ack, this is one of the reason I love code reviews, I completely forgot about that.

tharsheblows commented on local/public/wp-content/plugins/hello.php:
```diff
@@ -3,3 +3,6 @@
  */
+
+
+
```
This is so annoying -- I don't think it really matters but why is there a test that affects this file? Maybe we should take it out?

delawski replied:
I hate rolling back this change every time, too! 👿 

> Maybe we should take it out?

One hundred percent! 😄

tharsheblows replied:
I've updated the test so that it puts back the original contents of the file when it's done. It's working for me 😅

tharsheblows replied:
@delawski The updated test is here https://github.com/xwp/stream/pull/1529/files#diff-b72d8df01c58f9124052f8296360e9aa1a7bab799c60b038171ca3a23b9aa1c8

tharsheblows commented:
@delawski I've updated this to not run the e2e tests in the GH action by renaming it and added the files to `.distignore` but also have renamed the phpunit test directory to `phpunit` from `tests` to make it clearer what is where.

delawski commented on .gitignore:
```diff
@@ -21,3 +21,7 @@ alerts/js/*.min.js
 
 # Generated test data
 tests/data/tmp/*
+/test-results/
+/playwright-report/
+/blob-report/
+/playwright/.cache/
```
<!-- potential concern, ask for clarification or investigation -->
**❓question:** I know that in CI those are not (yet?) generated, but should we add them preemptively to `.distignore`?

tharsheblows replied:
YES, thank you! I have some sort of mental block on this!

delawski commented on playwright.config.js:
```diff
@@ -0,0 +1,79 @@
+
+  /* Configure projects for major browsers */
+  projects: [
+    {
+      name: 'chromium',
+      use: { ...devices['Desktop Chrome'] },
+    },
```
<!-- propose improvements -->
**💬 suggestion:** I believe we should only leave that one. I think that our intention is to test whether we've got all the functionality working but not really if it works in various browsers.

delawski commented on tests/e2e/editor-new-post.spec.js:
```diff
@@ -0,0 +1,63 @@
+		await expect( page.getByText( `"${ postTitle }" post published` ) ).toBeVisible();
+	} );
+
+	// We should not have an updated row. This times out which makes it fail.
+	test( 'does not have updated row', async () => {
+		// Expects Stream log to have "Test Post" post published visible.
+		await expect( page.getByText( `"${ postTitle }" post updated` ) ).not.toBeVisible();
+	} );
```
<!-- highlight specific problems -->
**❗️issue:** This one is not passing on my local:

<img width="989" alt="Screenshot 2024-07-31 at 13 55 51" src="https://github.com/user-attachments/assets/ced2393e-7532-4253-9f9c-1926147f00aa">

I tried with `fullyParallel: false`, too, but it failed in the same way.

tharsheblows replied:
@delawski -- I'm so sorry, this one is supposed to fail. It'll be fixed by https://github.com/xwp/stream/pull/1530.

delawski replied:
Perfect, thank you!

delawski reviewed (commented):
Great idea about organizing tests in the `tests` folder the way you did it 👍 

It looks like one of the tests is failing for me. Is it working okay on your local?

delawski commented:
@tharsheblows I'm sorry about the dependencies update train when you were trying to merge it to `develop` 😅

I resolved the conflicts now and will merge as soon as the CI checks are complete.
